compiler/rustc_resolve/src/lib.rs

```diff
@@ -1877,20 +1874,19 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         }
     }
 
-    /// Retrieves the span of the given `DefId` if `DefId` is in the local crate.
-    #[inline]
-    fn opt_span(&self, def_id: DefId) -> Option<Span> {
-        def_id.as_local().map(|def_id| self.tcx.source_span(def_id))
+    /// Retrieves definition span of the given `DefId`.
+    fn def_span(&self, def_id: DefId) -> Span {
+        match def_id.as_local() {
+            Some(def_id) => self.tcx.source_span(def_id),
+            None => self.cstore().get_span_untracked(def_id, self.tcx.sess),
+        }
     }
 
-    /// Retrieves the name of the given `DefId`.
-    #[inline]
-    fn opt_name(&self, def_id: DefId) -> Option<Symbol> {
-        let def_key = match def_id.as_local() {
-            Some(def_id) => self.tcx.definitions_untracked().def_key(def_id),
-            None => self.cstore().def_key(def_id),
-        };
-        def_key.get_opt_name()
+    fn associated_item_def_ids(&self, def_id: DefId) -> Option<&'tcx [DefId]> {
+        match def_id.as_local() {
+            Some(def_id) => self.associated_item_def_ids.get(&def_id).copied(),
+            None => Some(self.tcx.associated_item_def_ids(def_id)),
```

Comment:
The naming here is off.

The `associated_item_def_ids` query is only supposed to work on traits and impl blocks, to return the def-ids of the contained items. Using it to access an ADT's fields is a quirk of metadata decoding, which stores both information in the same `children` table and does not validate the parent before answering.

The "proper" way to access the information would use the `tcx.adt_def` query.

In the same idea, this method and the field in the first branch would be better named `field_def_ids`, as that's what they are.

Reply:
>The `associated_item_def_ids` query is only supposed to work on traits and impl blocks, to return the def-ids of the contained items.

I'd say that fields are close enough to bless this use.
Previously `associated_item_def_ids` was certainly used for fields in rustdoc (until https://github.com/rust-lang/rust/pull/94857 nuked all the relevant code) and maybe in rustc too.

>The "proper" way to access the information would use the `tcx.adt_def` query.

`adt_def` query cannot be used in resolve because it create cycles (it needs lang items).
Also it decodes much more information than necessary in our case, but it doesn't matter much because all its uses would be in diagnostics (but in general `associated_item_def_ids` relates to the `adt_def` query in the same way as to `associated_item`, i.e. minimizing decoding and dependencies).

>In the same idea, this method and the field in the first branch would be better named `field_def_ids`, as that's what they are.

Yeah, I'll rename.
I used the current naming to maybe extend it to other associated items in the future, but now I agree that it only brings confusion.